### ISSUE TITLE
[WIP] ColorManager proposal

### ISSUE THREAD

**Issue by kevinyamauchi** [napari/napari#1727]
This issue is a proposal for `ColorManager` a class to handle setting of colors in our layers where we color individual data points (i.e., `Points`, `Vectors`, `Shapes`, `Tracks`). Currently, this is handled by code in each layer. Many of the layers share code leading to a ton of code duplication. Additionally, there is a bit of fragmentation (e.g., `Tracks` handles coloring by property differently than `Points`, `Vectors`, and `Shapes`.

## Proposed API
Similar to our `TextManager`, the `ColorManager` would be a dataclass that contains the current color values as a property and has the methods for setting the colors. `ColorManager` would not store the layer `properties`. Each layer utilizing `ColorManager` would have one `ColorManger` object for each display property that has a color. 

```python
# create a points layer with blue points
points_layer = viewer.add_points(data, face_color='blue')
points_layer = viewer.add_points(data, face_color=[0, 0, 1, 1])

# create a points layer with face color set by the property 'class', use defaults
points_layer = viewer.add_points(data, properties=properties, face_color='class')

# create a points layer and specify some mapping options
face_color = {
    'color': 'class',
    'mode': 'colormap',
    'colormap': 'viridis'
}
points_layer = viewer.add_points(data, properties=properties, face_color=face_color)

# update the face color colormap
points_layer.face_color.colormap = 'plasma'
```

## Mapping modes
I propose the following modes for setting colors:
- `direct`: the user provides the exact colors to use. This is either a single color that is broadcasted or an array of colors (one for each datum.
- `direct_mapping`: a mapping of each property value to a color. This would be a dictionary where the keys are the property value and the values is the color.
- `color_cycle`: a cycle of color is applied to the property values. This is useful for categorical properties where the user doesn't want to specify a color for each property value.
- `color_map`: mapping of continuous property values to color.

## Questions

1. Do we want to maintain support for the current API? In particular, we have many properties for color at the layer level (e.g., `layer.face_color_mode` and `layer.edge_color_mode`) that would be unnecessary with `ColorManager`
2. Do we want persistence of settings when we switch between modes? Currently, if we are in `color_cycle` mode, switch to `colormap` mode, and then switch back to `color_cycle` mode, the original `color_cycle` settings are "remembered".

## Related issues
Some related issues:
- #1356 
- #1491

## Approach:

We could break this up into a series of PRs:
- [ ] Create `ColorManager`, implement in `Points`
- [ ] Create layer controls GUI for `ColorManager` and add to `Points`
- [ ] Add color manager + layer controls to `Shapes`
- [ ] Add color manager + layer controls to `Vectors`
- [ ] Add color manager + layer controls to `Tracks`

**Comment by kevinyamauchi** [comment]
@adamltyson, I recall that you are using properties to color point in some of your applications. It would be great to hear if this proposal would fit your use case and/or if there are any issues you come across that you think we could solve here.

**Comment by adamltyson** [comment]
@kevinyamauchi, yep, this proposal looks great. Currently I only have two classes so it's pretty simple, but soon we'll have an application with many more, and being able to flexibly encode these with colors would be great.

I don't have much to add, but this looks like a good idea to me. I don't think it would solve any specific issues, other than making my life a bit easier :thumbsup:

**Comment by kevinyamauchi** [comment]
Thanks for the feedback, @adamltyson !

**Comment by kevinyamauchi** [comment]
From [this comment](https://github.com/napari/napari/pull/2204#pullrequestreview-600185434) in #2204 by @sofroniewn , we will do the following after #2204 merges.

> After this merges, I could see the following next steps
> - [ ] refactor vectors to use a private `_edge` ColorManager
> - [ ] refactor shapes to use a private `_edge` and `_face` ColorManager
> - [ ] refactor tracks to use a private ColorManager (have to think about name)
> - [ ] Create a new `Edge > ColorManager` class that also included `width` as an attribute
> - [ ] Give points a public `edge` and `face` and deprecate old attributes.
> - [ ] Give vectors a public `edge` and deprecate old attributes
> - [ ] Give shapes a public `edge` and deprecate old attributes
> - [ ] Use tracks public colormanger

**Comment by sofroniewn** [comment]
Now that #2204 is merged I think we should continue the "private" refactor in as many places as possible first - @jni also mentioned trying to use this in labels too, and then reconvene before deciding exactly how to expose everything publicly.

I think @jni wasn't sold yet on the `layer.edge.color` API, but we can defer that decision until we've gotten further. How does that sound @kevinyamauchi?

**Comment by kevinyamauchi** [comment]
I agree that we should perform the private refactor first and then determine how we can should best set up the new public API when we can test it on all of the layers.

I like the aesthetics of `layer.edge.color`, but I think it will be difficult to implement that sort of hierarchal API for all `ColorManager` fields unless we allow `ColorManager` to access the layer. This is because since some of the `ColorManager` fields require knowing what the layer properties are to set properly (e.g., setting the color to map to a property by passing the name of the propety).

Perhaps we could keep the `layer.edge_color` setter, but it would accept both a value (e.g., color or property name) or a dictionary that can set multiple `ColorManager` fields. We can then in the setter "inject" the necessary values from the layer into the dictionary before passing it to the `ColorManager` update method. The API would then look something like:

```python
# set the edge color to a single color
points_layer.edge_color = 'red'

# set the edge color of each point individually
points_layer.edge_color = ['red', 'blue', 'magenta']

# set the edge color to map to the property point_type
points_layer.edge_color = 'point_type'

# set the edge color to map to the property point_type and set the colormap
points_layer.edge_color = {
    'color': 'point_type',
    'continuous_colormap': 'viridis',
    'color_mode': 'colormap'
}

```
